src/main/scala/com/wixpress/common/specs2/JMock.scala

```diff
@@ -222,3 +232,151 @@ class AnswerAction3[I1, I2, I3, K <: AnyRef](msg: String, f: (I1, I2, I3) => K)
 class AnswerAction4[I1, I2, I3, I4, K <: AnyRef](msg: String, f: (I1, I2, I3, I4) => K) extends AnswerAction(msg) {
   override def invoke(invocation: Invocation): AnyRef = f(invocation(0), invocation(1), invocation(2), invocation(3))
 }
+
+sealed trait DefaultArgsMode
+object DefaultArgsMode {
+  case object ArgsAreValues extends DefaultArgsMode
+  case object ArgsAreMatchers extends DefaultArgsMode
+}
+
+object Macros {
+  import scala.reflect.macros.blackbox
+
+  def allowingImpl(c: blackbox.Context)(mock: c.Tree)(expr: c.Tree): c.Tree = {
+    import c.universe._
+    expectation(c)(mock, expr, q"${c.prefix}.jmockDsl.allowing($mock)")
+  }
+
+  def neverImpl(c: blackbox.Context)(mock: c.Tree)(expr: c.Tree): c.Tree = {
+    import c.universe._
+    expectation(c)(mock, expr, q"${c.prefix}.jmockDsl.never($mock)")
+  }
+
+  def ignoringImpl(c: blackbox.Context)(mock: c.Tree)(expr: c.Tree): c.Tree = {
+    import c.universe._
+    expectation(c)(mock, expr, q"${c.prefix}.jmockDsl.ignoring($mock)")
+  }
+
+  def oneOfImpl(c: blackbox.Context)(mock: c.Tree)(expr: c.Tree): c.Tree = {
+    import c.universe._
+    expectation(c)(mock, expr, q"${c.prefix}.jmockDsl.oneOf($mock)")
+  }
+
+  def atLeastImpl(c: blackbox.Context)(n: c.Tree)(mock: c.Tree)(expr: c.Tree): c.Tree = {
+    import c.universe._
+    expectation(c)(mock, expr, q"${c.prefix}.jmockDsl.atLeast($n).of($mock)")
+  }
+
+  def atMostImpl(c: blackbox.Context)(n: c.Tree)(mock: c.Tree)(expr: c.Tree): c.Tree = {
+    import c.universe._
+    expectation(c)(mock, expr, q"${c.prefix}.jmockDsl.atMost($n).of($mock)")
+  }
+
+  def exactlyImpl(c: blackbox.Context)(n: c.Tree)(mock: c.Tree)(expr: c.Tree): c.Tree = {
+    import c.universe._
+    expectation(c)(mock, expr, q"${c.prefix}.jmockDsl.exactly($n).of($mock)")
+  }
+
+  private def expectation(c: blackbox.Context)(mock: c.Tree, expr: c.Tree, start:  c.Tree): c.Tree = {
+    import c.universe._
+    implicit val cc = c
+//    println(s"expr  raw: ${showRaw(expr)}")
```

Comment:
leftover?

Reply:
I needed these printouts through many iterations for debugging - this his basically the only way to figure out what's going on in the macro. Decided it to leave them commented out for the next time someone will need to debug this :)